### PR TITLE
Fix `__builtins__` in compiled functions

### PR DESCRIPTION
**Pull request by ZeroIntensity**
# Description

## ❓ What does this PR do?

Fixes the `__builtins__` attribute for compiled functions to match CPython.

## 🧐 Why was it initiated? Any relevant Issues?

- #3762

## 🧱 Type of Change

- [x] 🐛 Bug fix (non-breaking change which fixes an issue)
- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as
  expected)
- [ ] 🧹 Refactoring (no functional changes, no api changes)
- [ ] 🏗️ Build / CI System
- [ ] 📚 Documentation Update

## ✅ PR Checklist

- [x] **Correct base branch selected**: Should be `develop` branch.
- [x] **Formatting**: Enabled commit hook or executed `./bin/autoformat-nuitka-source`.
- [x] **Tests**: All tests still pass. (See
  [Running the Tests](https://nuitka.net/doc/developer-manual.html#running-the-tests)).
  - CI actions cover the basics, but manual verification is encouraged.
- [x] **New Coverage**: New features or fixed regressions are covered via new tests.
- [ ] **Documentation**: Documentation updates are included for new or changed features.

## 🤖 AI Generated Code Policy

- [ ] **Detection**: This PR contains AI generated code.
  - [ ] **Issue First**: I have created an issue explaining the problem *before* using AI to
    generate a fix, ensuring the direction is correct.
  - [ ] **Documentation**: I have provided the prompts used to generate the code (non-optional).
  - [ ] **Verification**: I have **manually verified** the AI generated code.
    > **Note**: AI generated code is welcome, but it must be peer-reviewed and understood by the
    > submitter. Blindly copying AI output without understanding is not acceptable.
  - [ ] **Evidence**: I have included test evidence that the change is effective.

______________________________________________________________________

## Summary by Sourcery

Align compiled function __builtins__ handling with CPython semantics.

Bug Fixes:
- Ensure compiled functions expose __builtins__ as the builtins module’s __dict__ rather than the module object itself.

Tests:
- Add a basic test verifying the __builtins__ attribute on an empty compiled function.

### CODE REVIEW

File: nuitka/build/static_src/CompiledFunctionType.c [hunk 1]
```diff
@@ -584,7 +584,15 @@ static PyObject *Nuitka_Function_get_builtins(PyObject *self, void *data) {
...
     PyThreadState *tstate = PyThreadState_GET();
     struct Nuitka_FunctionObject *function = (struct Nuitka_FunctionObject *)self;
-    return LOOKUP_SUBSCRIPT(tstate, PyModule_GetDict(function->m_module), const_str_plain___builtins__);
+    PyObject *builtins_module =
+        LOOKUP_SUBSCRIPT(tstate, PyModule_GetDict(function->m_module), const_str_plain___builtins__);
+    if (builtins_module == NULL) {
+        return NULL;
+    }
+
+    PyObject *builtins_dict = LOOKUP_ATTRIBUTE(tstate, builtins_module, const_str_plain___dict__);
```
[the remark this answers is not in the document]
Comment (ZeroIntensity) [reply nuitka/build/static_src/CompiledFunctionType.c:593]: `__builtins__` is always the module here. CPython does not allow setting `__builtins__` on function objects, so I don't know what you're talking about.

File: tests/basics/FunctionsTest32.py [hunk 2]
```diff
@@ -180,6 +180,13 @@ def g():
 
 print("Using deleted non-local variable:", deletingClosureVariable())
 
+
+def emptyFunction():
+    pass
+
+
+print(emptyFunction.__builtins__)
```
[the remark this answers is not in the document]
Comment (kayhayen) [reply tests/basics/FunctionsTest32.py]: For Nuitka printing is the preferred way of testing, asserts could go wrong at run-time due to Nuitka bugs.

File: tests/basics/FunctionsTest32.py [hunk 3]
```diff
@@ -185,7 +185,11 @@ def emptyFunction():
     pass
 
 
-print(emptyFunction.__builtins__)
+if hasattr(emptyFunction, "__builtins__"):
```
Comment (kayhayen) [inline tests/basics/FunctionsTest32.py]: Kind of means that Functions32.py is not the right place for it. I guess, we can just move it to Functions.py and be good with it.

**Review by kayhayen — APPROVED** [review]

**Comment by kayhayen** [comment]
Part of 4.0.1
